src/Nethermind/Nethermind.Core/Extensions/Bytes.cs

```diff
@@ -679,88 +677,45 @@ private static int CountLeadingZeros(byte[] bytes)
-                hexString = hexString.Insert(startIndex, "0");
+                return c - 'a' + 10;
             }
-
-            int numberChars = hexString.Length - startIndex;
-
-            byte[] bytes = new byte[numberChars / 2];
-            for (int i = 0; i < numberChars; i += 2)
+            if (c >= 'A' && c <= 'F')
             {
-                bytes[i / 2] = Convert.ToByte(hexString.Substring(i + startIndex, 2), 16);
+                return c - 'A' + 10;
             }
 
-            return bytes;
+            throw new ArgumentException($"Invalid hex char {c}");
         }
 
-        private static byte[] FromHexNibble1Table =
-        {
-            255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
-            255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
-            255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
-            255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
-            255, 255, 255, 255, 255, 255, 255, 255, 0, 16,
-            32, 48, 64, 80, 96, 112, 128, 144, 255, 255,
-            255, 255, 255, 255, 255, 160, 176, 192, 208, 224,
-            240, 255, 255, 255, 255, 255, 255, 255, 255, 255,
-            255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
-            255, 255, 255, 255, 255, 255, 255, 160, 176, 192,
-            208, 224, 240
-        };
-
-        private static byte[] FromHexNibble2Table =
-        {
-            255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
-            255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
-            255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
-            255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
-            255, 255, 255, 255, 255, 255, 255, 255, 0, 1,
-            2, 3, 4, 5, 6, 7, 8, 9, 255, 255,
-            255, 255, 255, 255, 255, 10, 11, 12, 13, 14,
-            15, 255, 255, 255, 255, 255, 255, 255, 255, 255,
-            255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
-            255, 255, 255, 255, 255, 255, 255, 10, 11, 12,
-            13, 14, 15
-        };
-
         [DebuggerStepThrough]
         public static byte[] FromHexString(string hexString)
         {
-            if (hexString == null)
+            if (string.IsNullOrEmpty(hexString))
             {
                 throw new ArgumentNullException($"{nameof(hexString)}");
             }
 
-            int startIndex = hexString.StartsWith("0x") ? 2 : 0;
-            bool odd = hexString.Length % 2 == 1;
-            int numberChars = hexString.Length - startIndex + (odd ? 1 : 0);
-            byte[] bytes = new byte[numberChars / 2];
-            for (int i = 0; i < numberChars; i += 2)
+            int s = (hexString[0] == '0' && hexString[1] == 'x' ? 2 : 0);
+            int b = 0;
+            byte[] bytes = new byte[(hexString.Length - s + 1) / 2];
+            if ((hexString.Length & 1) != 0)
             {
-                if (odd && i == 0)
-                {
-                    bytes[0] += FromHexNibble2Table[(byte) hexString[startIndex]];
```

Comment:
why would you revert one of the most critical performance optimization in the code?
do you have a benchmark for comparison of the old and new code?